apps/price_pusher/src/evm/evm.ts

```diff
@@ -61,53 +84,44 @@ export class EvmPriceListener extends ChainPriceListener {
     await super.start();
   }
 
-  private async startSubscription() {
-    for (const { id: priceId } of this.priceItems) {
-      this.pythContract.events.PriceFeedUpdate(
-        {
-          filter: {
-            id: addLeading0x(priceId),
-            fresh: true,
-          },
-        },
-        this.onPriceFeedUpdate.bind(this)
-      );
-    }
+  private async startWatching() {
+    this.pythContract.watchEvent.PriceFeedUpdate(
+      { id: this.priceItems.map((item) => addLeading0x(item.id)) },
+      { onLogs: this.onPriceFeedUpdate.bind(this) }
+    );
   }
 
-  private onPriceFeedUpdate(err: Error | null, event: EventData) {
-    if (err !== null) {
-      this.logger.error(
-        err,
-        "PriceFeedUpdate EventEmitter received an error.."
+  private onPriceFeedUpdate(
+    logs: WatchContractEventOnLogsParameter<typeof PythAbi, "PriceFeedUpdate">
+  ) {
+    for (const log of logs) {
+      const priceId = removeLeading0x(log.args.id!);
+
+      const priceInfo: PriceInfo = {
+        conf: log.args.conf!.toString(),
+        price: log.args.price!.toString(),
+        publishTime: Number(log.args.publishTime!),
```

Comment:
In general I strongly advise against nearly ever using type casts, including `!`, in typescript.  It's not hard to just add an explicit check (e.g. something like `if (log.args.conf === undefined) { throw new MissingLogArgError('conf'); }` and adding the explicit check avoids possible scenarios where the runtime code does not behave in ways that the type system claims to guarantee.  As an added bonus, the runtime checks can help make it obvious when there's some structure in the code that could be cleaned up / improved / types that can be improved / better schema validation that could be happening.

There are a few scenarios where type casts are necessary due to limitations around how typescript does dependent typing but it's rare and this isn't one of the cases.

Reply:
This is a strange typing from Viem. They should not be optional at all. I'll add strict defined checks here.

Reply:
fwiw I usually have a helper in most codebases, something like:

```ts
const assertDefined = <T>(value: T | undefined): T => {
    if (value === undefined) {
        throw new Error('Assertion failed: value was undefined');
    } else {
        return value;
    }
};
```

Then instead of using `someVar!` in other code, I can use `assertDefined(someVar)`.  Nearly as easy to type, but also more obvious to read, and not subject to making the code unsafe.

Reply:
Thanks. Apparently Viem has a strict mode that handles types but with the contract instance even if I set it the types won't get resolved. @jxom too